File: typescript/src/giftcard/domain/policies/refundGiftCardPolicy.ts
// Policy che traduce i risultati della prenotazione in rimborsi sulla gift card.
// Violazione cross-BC didattica: GiftCard non dovrebbe importare tipi da Booking.

import type { BookingResultEvent } from '@/booking/domain/events/bookingResultEvents.js';
import type { Policy } from '@/common/domain/model/policy.js';
import { match } from 'ts-pattern';
import { RefundGiftCard, refundGiftCard } from '../../application/commands/refundGiftCard.js';

export class RefundGiftCardPolicy implements Policy<BookingResultEvent, RefundGiftCard> {
  evaluate(event: BookingResultEvent): RefundGiftCard | null {
    return match(event)
      .with(
        { kind: 'BookingRejected' },
        (rejected) => refundGiftCard(rejected.giftCardId, rejected.amount)
      )
      .with({ kind: 'BookingConfirmed' }, () => null)
      .with({ kind: 'BookingRefused' }, () => null)
      .exhaustive();
  }
}

File: java/src/main/java/it/giannibombelli/wsc2026/giftcard/application/policies/RefundGiftCardPolicy.java
package it.giannibombelli.wsc2026.giftcard.application.policies;

import it.giannibombelli.wsc2026.booking.domain.events.BookingResultEvents;
import it.giannibombelli.wsc2026.common.application.Policy;
import it.giannibombelli.wsc2026.common.utils.Require;
import it.giannibombelli.wsc2026.giftcard.application.commands.RefundGiftCard;


public final class RefundGiftCardPolicy implements Policy<BookingResultEvents, RefundGiftCard> {
    @Override
    public RefundGiftCard evaluate(BookingResultEvents event) {
        Require.requireArgument(event, "Booking event");
        return switch (event) {
            case BookingResultEvents.BookingRejected it -> new RefundGiftCard(it.giftCardId(), it.amount());
            case BookingResultEvents.BookingConfirmed ignored -> null;
            case BookingResultEvents.BookingRefused ignored -> null;
        };
    }
}
